mars/dataframe/dataset.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 1999-2021 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import pandas as pd
+from ..utils import lazy_import
+from typing import Iterable, List
+
+ray = lazy_import('ray')
+parallel_it = lazy_import('ray.util.iter')
+ml_dataset = lazy_import('ray.util.data')
+
+
+class RayObjectPiece:
+    def __init__(self,
+                 addr,
+                 obj_ref):
+        """Represents a single entity holding the object ref."""
+        self.addr = addr
+        self.obj_ref = obj_ref
+
+    def read(self, shuffle: bool) -> pd.DataFrame:
+        df: pd.DataFrame = ray.get(self.obj_ref)
+
+        if shuffle:
+            df = df.sample(frac=1.0)
+        return df
+
+
+class RecordBatch:
+    def __init__(self,
+                 shard_id: int,
+                 prefix: str,
+                 record_pieces: List[RayObjectPiece],
+                 shuffle: bool,
+                 shuffle_seed: int):
+        """Holding a list of RayObjectPieces."""
+        self._shard_id = shard_id
+        self._prefix = prefix
+        self.record_pieces = record_pieces
+        self.shuffle = shuffle
+        self.shuffle_seed = shuffle_seed
+
+    @property
+    def prefix(self) -> str:
+        return self._prefix
+
+    @property
+    def shard_id(self) -> int:
+        return self._shard_id
+
+    def __iter__(self) -> Iterable[pd.DataFrame]:
+        """Returns the item_generator required from ParallelIteratorWorker."""
+        if self.shuffle:
+            np.random.seed(self.shuffle_seed)
+            np.random.shuffle(self.record_pieces)
+
+        for piece in self.record_pieces:
+            yield piece.read(self.shuffle)
+
+
+def _create_ml_dataset(name: str,
+                       record_pieces: List[RayObjectPiece],
+                       num_shards: int,
+                       shuffle: bool,
+                       shuffle_seed: int):
+    if shuffle_seed:
+        np.random.seed(shuffle_seed)
+    else:
+        np.random.seed(0)
+
+    # TODO: (maybe) combine some chunks according to num_shards
+    record_batches = []
+    for rank, pieces in enumerate(record_pieces):
+        record_batches.append(RecordBatch(shard_id=rank,
```

Comment:
1. Do we need to make one ParallelIteratorWorker hold all obejct refs in current node? If num_shards not specified and chunk size is very small, there will be many `ParallelIteratorWorker` actors.
2. If one band has many object refs, will there be an OOM when iterating parallel iterator?